compiler-core/src/language_server/annotation_printer.rs

```diff
@@ -0,0 +1,297 @@
+#![allow(dead_code)]
+
+use ecow::EcoString;
+use im::{HashMap, HashSet};
+use std::sync::Arc;
+
+use crate::{
+    ast::{TypeAst, UnqualifiedImport},
+    type_::{Type, TypeVar, PRELUDE_MODULE_NAME},
+};
+
+#[derive(Debug)]
+pub struct AnnotationPrinter<'a> {
+    names: HashMap<u64, EcoString>,
+    uid: u64,
+
+    // Type aliases that are defined in the current module
+    // Key: (module name, type name), Value: Alias
+    // eg: A type alias statement like so :
+    // type UserId = Int
+    // would result in  Key: (mod, UserId), Value: Int
+    type_aliases: &'a HashMap<(EcoString, EcoString), EcoString>,
+
+    // Mapping of imported modules to their aliased named
+    // Key: Module name, Value: Alias
+    // eg: import mod1 as M -> Key: mod1, Value: M
+    module_aliases: &'a HashMap<EcoString, EcoString>,
+
+    // Generic type parameters that have been annotated by the user
+    // eg:
+    // fn foo(arg1: bar, arg2){
+    //  arg1 == arg2
+    // } -> Here bar is an annoation. So the key would be its generic type ID and the associated value would be bar.
+    // This needs to be kept track to ensure that the generated type annotation of arg2 is bar(user-given) and not the generic type(a, b, ..) that is generated by the compiler.
+    generic_annotations: &'a HashMap<u64, TypeAst>,
+
+    // These are unqualified types that are imported in the current module.
+    unqualified_imports: &'a HashMap<(EcoString, EcoString), UnqualifiedImport>,
+
+    // A set of unqualified type names that are valid in the current module - used to check if a prelude type is shadowed.
+    unqualified_type_names: &'a HashSet<EcoString>,
+}
+
+impl<'a> AnnotationPrinter<'a> {
+    pub fn new(
+        type_aliases: &'a HashMap<(EcoString, EcoString), EcoString>,
+        module_aliases: &'a HashMap<EcoString, EcoString>,
+        generic_annotations: &'a HashMap<u64, TypeAst>,
+        unqualified_imports: &'a HashMap<(EcoString, EcoString), UnqualifiedImport>,
+        unqualified_type_names: &'a HashSet<EcoString>,
+    ) -> Self {
+        AnnotationPrinter {
+            names: im::HashMap::new(),
+            uid: u64::default(),
+            type_aliases,
+            module_aliases,
+            generic_annotations,
+            unqualified_imports,
+            unqualified_type_names,
+        }
+    }
+
+    pub fn print_type(&mut self, typ: &Type) -> String {
```

Comment:
We use EcoString, not String. If conversion is needed to interact with some library we should do it at the boundary of the system, not internally